### DISCUSSION TITLE
does pyg have group_apply_edges functionality?

### DISCUSSION THREAD

**Discussion by ShengzheXu** [pyg-team/pytorch_geometric#3599] [Q&A] [answered]
I saw some code online using DGL library and they can do operations on a group of edges that related to some node. For example, they can do softmax normalization to the "target node". Just like the followed webpage says. Do we also have similar methods to do so? Thanks for any advice.

https://docs.dgl.ai/en/0.4.x/generated/dgl.DGLGraph.group_apply_edges.html

**Comment by wsad1** [comment, marked as answer]
Yes, [torch_scatter](https://pytorch-scatter.readthedocs.io/en/latest/) performs operations based on a group index.
For example, you can perform the the softmax operation you mentioned using [torch_goemetric.utils.softmax](https://pytorch-geometric.readthedocs.io/en/latest/modules/utils.html), which uses torch_scatter internally, as follows
```
source, target = edge_index # 2*E
attr = edge_attr # E*F
out = softmax(attr, target) # softmax on features of edges incident on target node.
out = softmax(attr, source) # softmax on features of edges with common source.
```

**Reply by ShengzheXu** [reply]
Thanks Jinu, this example really helps!
